Bootnode replacement: elastic IPs + autoscaling groups

john-osullivan opened the pull request:
This PR refactors bootnodes from directly using AWS instances to instead use an autoscaling group, so that bootnodes are automatically replaced when they fail.  In order to ensure they have a static IP, we also use elastic IPs which are provisioned at `terraform apply` and are then associated to each autoscaling group's node within the `init-bootnode.sh` script.

- Code has been tested on a unit test net (1-1-1-1 in us-east-1), private transactions were successful.
- The bootnode DNS names returned by terraform data providers now fails, as claiming a new IP address also changes the bootnodes' DNS names.  Solution was to use the permanent IP in key places where the hostname was being used (writing to Vault & the terraform output).
- Added a new permission, all nodes are able to claim any address.  This technically allows an attacker with access to one bootnode to make the rest of them inaccessible by claiming all of their IPs.  However, re-running `terraform apply` ought to re-associate each node with its proper address.
- If somebody launches a network with more than 5 bootnodes in this version, AWS will throw an error requiring that they request more elastic IPs.  Our proposed solution there was to have EIP functionality disabled by default, that is not implemented in this PR.  Based on the number of changes required here, I wanted to sit down and talk about the right way to make this functionality optional.

john-osullivan committed:
Changed README bullet to quote

john-osullivan committed:
Updated README with initial summary & implementation strategy

john-osullivan committed:
First draft of infra & user-data-script changes, still missing connection b/w LB & ASG

john-osullivan committed:
Added the lb listener & target group, just missing IP resolution in the init script

john-osullivan committed:
Beginning conversion to resolve public IP & hostname using load balancer's DNS

john-osullivan committed:
Determined that LB_DNS==HOSTNAME, beginning IP tests

john-osullivan committed:
Switched strategy from LBs to elastic IPs

john-osullivan committed:
wip

john-osullivan committed:
Removed unnecessary Terraform provisioners fxn, added IAM perms for ec2:AssociateAddress

john-osullivan committed:
Now writing static IP to Vault, fixed bug in launch config specifying user data file

john-osullivan committed:
Replaced bootnode URL w/ IPs, as the URL gets replaced after node turns on

john-osullivan committed:
Removed temp README changes

john-osullivan committed:
More pre-PR cleanup

Lsquared13 reviewed (changes requested):
Let's keep existing naming conventions. Use dashes instead of underscores.

`public-ip.txt` and `eip-id.txt`

Lsquared13 commented on packer/instance-scripts/init-bootnode.sh:
```diff
@@ -79,15 +72,18 @@ wait_for_successful_command 'vault status'
 HOSTNAME=$(wait_for_successful_command 'curl http://169.254.169.254/latest/meta-data/public-hostname')
 BOOT_PORT=30301
 
+# Associate the EIP with this instance
```
Do we only need to do this once?

john-osullivan replied:
Yup!  This only needs to happen when the node turns on, then the EIP is associated until it's explicitly reclaimed or reassociated with a new node.  When a replacement node spins up and re-runs the command, it will then get this IP.

Lsquared13 commented on terraform/modules/quorum-cluster-region/outputs.tf:
```diff
@@ -11,5 +11,5 @@ output "quorum_observer_node_dns" {
 }
 
 output "bootnode_dns" {
```
Rename this to `bootnode_ip`, this is technically inaccurate

john-osullivan committed:
Addressed PR comments

john-osullivan commented:
My most recent commit addresses all those comments.  Note on the dashes v. underscores one -- is there a reason we have the [terraform variables use underscores but then switch to dashes for the .txt files](https://github.com/Eximchain/terraform-aws-quorum-cluster/blob/master/terraform/modules/quorum-cluster-region/user-data/user-data-quorum.sh#L66)?  My original naming was to keep in line with terraform variable names.  Obviously doesn't matter one way or the other, just thought I'd point out where the gap was.

john-osullivan committed:
Tentative update to make feature conditional.  Still testing

Lsquared13 commented on packer/instance-scripts/init-bootnode.sh:
```diff
@@ -79,15 +72,30 @@ wait_for_successful_command 'vault status'
 BOOT_PORT=30301
 
+# Associate the EIP with this instance
+echo ">> Value of USING_EIP follows on next line:"
```
Remove or clean up these messages

Lsquared13 commented on packer/instance-scripts/init-bootnode.sh:
```diff
@@ -79,15 +72,30 @@ wait_for_successful_command 'vault status'
+then
+    echo ">> Found USING_EIP == 1, running aws ec2 associate-address"
+    wait_for_successful_command "aws ec2 associate-address --instance-id $INSTANCE_ID --allocation-id $EIP_ID --region $AWS_REGION --allow-reassociation"
+else
```
Maybe make this an `elif` statement that checks for 0 so if and when they break this behavior we can fail loudly in the else block?

Lsquared13 commented on terraform/modules/quorum-cluster-region/bootnodes.tf:
```diff
@@ -32,50 +32,59 @@ resource "aws_subnet" "bootnodes" {
-    # The connection will use the local SSH agent for authentication if this is empty.
-    private_key = "${var.private_key}"
-  }
+  name_prefix = "quorum-bootnode-net-${var.network_id}-node-${count.index}"
```
nit: throw a `-` at the end of this `name_prefix`

Lsquared13 reviewed (changes requested):
Just a few more nits

john-osullivan committed:
Addressed nits, fixed terraform error, moved HOSTNAME check to after the EIP assoc

john-osullivan committed:
Made EIP list non-empty so element fxn would stop complaining at me

john-osullivan commented:
Just ran a unit private tx test with this new feature disabled, everything behaved.  Running another with it enabled now to finish verifying that everything is up to speed.

john-osullivan commented:
The enabled case also succeeded -- this should be good to go!

Lsquared13 commented on terraform/modules/quorum-cluster-region/bootnodes.tf:
```diff
@@ -84,11 +93,26 @@ resource "aws_instance" "bootnode" {
   template = "${file("${path.module}/user-data/user-data-bootnode.sh")}"
 
   vars {
+    constellation_s3_bucket = "${aws_s3_bucket.quorum_constellation.id}"
```
nit: Align the `=`

So keep the longest variable name as-is and pad the others with spaces until the `=` align vertically.

You don't need to re-run the test since it's a purely cosmetic change

Lsquared13 reviewed (commented):
Looks good otherwise

john-osullivan committed:
Aligned equal signs in bootnode, updated comment w/ info from @apparentlymart

Lsquared13 commented on terraform/modules/quorum-cluster-region/bootnodes.tf:
```diff
@@ -84,20 +93,36 @@ resource "aws_instance" "bootnode" {
+    # as element() cannot be called on empty list.  Solution is hacky, but lazy
+    # ternary evaluation will drop in Terraform 0.12: https://www.hashicorp.com/blog/terraform-0-1-2-preview
+    # If you're reading this and it has already released, try dropping the concat hack.
+    public_ip                = "${var.use_elastic_bootnode_ips ? element(concat(aws_eip.bootnodes.*.public_ip, list("")), count.index) : "nil"}"
```
Whoops, should have been more specific about the convention. What I meant was to line up the user data vars. 

The convention is to line up `=` for any variables listed on consecutive lines. If you skip a line, the `=` don't need to align with the ones in the previous block.

So for example, `vault_dns` and `vault_port` should be aligned with each other, but don't need to align with `eip_id` above it or `consul_cluster_tag_key` below it

john-osullivan committed:
equal spacing within blocks, not across
